### PR TITLE
BOM Column hiding and reordering

### PR DESCRIPTION
**Pull request by phagenlocher**
For some projects, having the _Footprint_ column is not needed and just wastes space. Thus, this patch proposes an option that allows to show/hide the footprints column.

image

Especially on smaller screens the table gets hard to read if you want to get a bigger view of the PCB.

image

However, with the  _Footprint_ column removed, it looks much nicer. The table cells also take up less space which makes it possible to view more components without scrolling.

image

### CODE REVIEW

**Comment by qu1ck** [comment]
I will not merge this as is because the scope is too narrow. There is a related wishlist item to make all hardcoded columns configurable and reorderable. See also discussion in #167.

If you'd like to work with me on this here is the direction I can suggest: 
Make a small popup menu appear on hover in the currently unused block at top left corner of the bom table. That menu should have all columns with ability to toggle their visibility and reorder them by drag and drop (preferred) or select and up/down buttons (probably easier to implement, clunkier to use). The first column (row number) and checkbox columns are special case here and the first one probably should just be toggleable but not reorderable while latter should be treated as a single block.
The state should also be preserved in localstorage, as with other settings.

**Comment by phagenlocher** [comment]
I dont think that reordering inside a pop-up menu makes for a good user experience. Why not have a drag and drop reordering on the column header and use the pop up purely for visibility checkboxes?

Should reordering of the checkbox columns then also change the order of the "Bom checkboxes" option in the preferences pop-up?

**Comment by qu1ck** [comment]
> Why not have a drag and drop reordering on the column header and use the pop up purely for visibility checkboxes?

Agreed, that's even better.

> Should reordering of the checkbox columns then also change the order of the "Bom checkboxes" option in the preferences pop-up?

Either treat them as a single block and rely on ordering in preferences pop-up or if they will be movable individually then list in preferences should be updated accordingly.

**Comment by qu1ck** [comment]
Some notes on implementation:
* The column hiding/reordering needs to support extra fields
* It's best to store the data in 2 generic fields: hiddenColumns list and columnOrder list instead of hardcoding visibility toggle and storage for each column.

**Comment by phagenlocher** [comment]
Agreed. I'll refactor the code and mark this as WIP.

**Comment by phagenlocher** [comment]
Columns can now be reordered with drag'n'drop on the column headers. Checkboxes and extra fields are treated as a single column.

`hiddenColumns` and `columnOrder` have been added to the settings. The drag'n'drop handles the change of `columnOrder`, `hiddenColumns` is changed by the pop-up menu, which still uses the cog symbol (change that?).

When the visible columns are reordered, the hidden ones are automatically pushed to the right in the settings.

**Comment by qu1ck** [comment]
I played around with it a bit and it works but there are some issues with dark theme and column placement detection logic. I have to drag waaay to the left for it to work. 

https://i.imgur.com/6eQxwNM.gif

Also I'd like to see following changed:
* extra fields should be treated same as normal columns, i.e. not a single block. In the future difference between them in html will be minimal when currently hardcoded columns will also be optional. 
* Checkboxes in hide columns menu should be generated dynamically and their order should match table order
* Not sure about shifting hidden columns to the right when reordering happens. It shouldn't be much harder to not do that and keep their place in the array. If you drag a column over a hidden columns place then it doesn't matter which way you choose to place the hidden one but it should still keep it's relative position to other columns.

For the icon use this
```
background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24'%3E%3Cpath fill='none' stroke='%23333' d='M2.5 4.5h5v15h-5zM9.5 4.5h5v15h-5zM16.5 4.5h5v15h-5z'/%3E%3C/svg%3E");
```

And best place for it is in top left corner of the bom table in the unused cell of the header.

**Comment by phagenlocher** [comment]
Dragging should work now.

If all hardcoded columns are at some point treated as optional there has to be a standard way of naming them. How should that look? Lowercasing the header name?

I don't like the idea of elements in a menu changing their order when columns in a table get reordered, even though they are functionally linked. It seems counterintuitive to me.

Pushing all hidden columns to the right was the easiest way of implementing it. I didn't put too much thought into it since I don't suspect users to reorder columns very often.

**Comment by qu1ck** [comment]
> If all hardcoded columns are at some point treated as optional there has to be a standard way of naming them. How should that look? Lowercasing the header name?

Use header name as is.

> I don't like the idea of elements in a menu changing their order when columns in a table get reordered, even though they are functionally linked. It seems counterintuitive to me.

Why? It's more counterintuitive to me that if I move last column to be first it's still last in visibility menu. If you have a bunch of columns with extra fields navigating the list will be much easier if it corresponds to order in the table.

One more thing I noticed: search field still matches hidden columns. Ref lookup field should probably be disabled when references column is hidden.

**Comment by qu1ck** [comment]
To clarify the standard column naming; they will be still hardcoded but in python part, not in html.
In html instead of config.extra_fields there will be config.fields and bom table will be a flat array of fields corresponding to config.fields. The standard columns will still get some special treatment for ref lookup, smart value sorting and similar edge cases but that will only be in effect if columns are present in config.fields.

**Comment by phagenlocher** [comment]
The visibility menu button has now moved to the free space in the table header and is dynamically built from the fields in the order they appear in the table. Reflookup is disabled and cleared on hiding the references column. The filter now ignores hidden columns.

**Comment by qu1ck** [comment]
Works pretty well. Aside from cosmetic stuff I mentioned inline only 2 things left to address:
* Extra fields to be treated individually
* Relative order of hidden columns to be preserved when dragging (as opposed to be shifted all the way to the right)

**Comment by phagenlocher** [comment]
Extra fields are now treated as normal columns. Also, the order of hidden columns is now preserved.  The code should also work once all columns are generalized. The `th` HTML elements now have a "col_name" attribute which identifies columns by name for the drag'n'drop.

**Comment by qu1ck** [comment]
Please address inline comments I made earlier and remove WIP when this is ready for final review.

**Comment by phagenlocher** [comment]
Which comments are you referring to?

**Review by qu1ck — COMMENTED** [review]
Oops, looks like I didn't submit them, sorry.

File: InteractiveHtmlBom/web/ibom.js [hunk 1]
```diff
@@ -347,14 +386,58 @@ function createColumnHeader(name, cls, comparator) {
...
+
+  var vismenu = document.createElement("div");
+  vismenu.id = "vismenu";
+  vismenu.classList.add("menu");
+  
+  var visbutton = document.createElement("button");
+  visbutton.classList.add("visbtn");
+
+  var viscontent = document.createElement("div");
+  viscontent.classList.add("menu-content");
```
[the remark this answers is not in the document]
Comment (qu1ck) [reply InteractiveHtmlBom/web/ibom.js:406]: Actually add it on the button, not menu content.

**Comment by phagenlocher** [comment]
This should address all comments. I could not find a nicer way around overwriting the font-family in `#vismenu-content` again, since `.bom` overwrites it. However, I consider this ready for final review.

**Comment by qu1ck** [comment]
There are some minor css and formatting issues but I'll take care of it a bit later.

Thanks for your contribution, this is a pretty neat feature!

**Pull request merged**